File: internal/share/logger.go
package share

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"runtime"
	"strings"
	"time"
)

// LoggerConfig controls logger behavior.
type LoggerConfig struct {
	Component    string `json:"component" yaml:"component"`
	Service      string `json:"service" yaml:"service"`
	Level        string `json:"level" yaml:"level"`
	Structured   bool   `json:"structured" yaml:"structured"`
	EnableCaller bool   `json:"enable_caller" yaml:"enable_caller"`
	Output       string `json:"output" yaml:"output"`
}

// Logger defines the shared logging interface.
type Logger interface {
	DEBUG(mcode MCode, optionalMessage string, fields ...map[string]any)
	INFO(mcode MCode, optionalMessage string, fields ...map[string]any)
	WARN(mcode MCode, optionalMessage string, fields ...map[string]any)
	ERROR(mcode MCode, optionalMessage string, fields ...map[string]any)
	FATAL(mcode MCode, optionalMessage string, fields ...map[string]any)
}

type LogLevel int

const (
	LevelDebug LogLevel = iota
	LevelInfo
	LevelWarn
	LevelError
	LevelFatal
)

func (rcvr LogLevel) String() string {
	switch rcvr {
	case LevelDebug:
		return "DEBUG"
	case LevelInfo:
		return "INFO"
	case LevelWarn:
		return "WARN"
	case LevelError:
		return "ERROR"
	case LevelFatal:
		return "FATAL"
	default:
		return "UNKNOWN"
	}
}

// LogEntry is the structured representation written by Logger.
type LogEntry struct {
	Timestamp string         `json:"timestamp"`
	Level     string         `json:"level"`
	Code      string         `json:"code"`
	Component string         `json:"component,omitempty"`
	Service   string         `json:"service,omitempty"`
	Message   string         `json:"message"`
	Fields    map[string]any `json:"fields,omitempty"`
	File      string         `json:"file,omitempty"`
	Function  string         `json:"function,omitempty"`
	Line      int            `json:"line,omitempty"`
	Error     string         `json:"error,omitempty"`
}

type logger struct {
	config LoggerConfig
	level  LogLevel
	output io.Writer
}

// NewLogger creates a shared logger from config.
func NewLogger(config LoggerConfig) Logger {
	return newLogger(config, nil)
}

func newLogger(config LoggerConfig, output io.Writer) Logger {
	logger := &logger{
		config: config,
		level:  parseLogLevel(config.Level),
		output: output,
	}
	if logger.output == nil {
		logger.output = openLogOutput(config.Output)
	}
	return logger
}

func parseLogLevel(level string) LogLevel {
	switch strings.ToUpper(strings.TrimSpace(level)) {
	case "DEBUG":
		return LevelDebug
	case "WARN", "WARNING":
		return LevelWarn
	case "ERROR":
		return LevelError
	case "FATAL":
		return LevelFatal
	default:
		return LevelInfo
	}
}

func openLogOutput(output string) io.Writer {
	switch strings.TrimSpace(output) {
	case "", "stdout":
		return os.Stdout
	case "stderr":
		return os.Stderr
	default:
		file, err := os.OpenFile(output, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o666)
		if err != nil {
			fmt.Fprintf(os.Stderr, "logger output fallback: %s: %v\n", output, err)
			return os.Stdout
		}
		return file
	}
}

func optionalMessage(mcode MCode, optional string) string {
	if optional == "" {
		return mcode.Message
	}
	if mcode.Message == "" {
		return optional
	}
	return fmt.Sprintf("%s: %s", mcode.Message, optional)
}

func (rcvr *logger) DEBUG(mcode MCode, optionalMessage string, fields ...map[string]any) {
	rcvr.log(LevelDebug, mcode, optionalMessage, firstFields(fields))
}

func (rcvr *logger) INFO(mcode MCode, optionalMessage string, fields ...map[string]any) {
	rcvr.log(LevelInfo, mcode, optionalMessage, firstFields(fields))
}

func (rcvr *logger) WARN(mcode MCode, optionalMessage string, fields ...map[string]any) {
	rcvr.log(LevelWarn, mcode, optionalMessage, firstFields(fields))
}

func (rcvr *logger) ERROR(mcode MCode, optionalMessage string, fields ...map[string]any) {
	rcvr.log(LevelError, mcode, optionalMessage, firstFields(fields))
}

func (rcvr *logger) FATAL(mcode MCode, optionalMessage string, fields ...map[string]any) {
	rcvr.log(LevelFatal, mcode, optionalMessage, firstFields(fields))
	os.Exit(1)
}

func firstFields(fields []map[string]any) map[string]any {
	if len(fields) == 0 {
		return nil
	}
	return fields[0]
}

func (rcvr *logger) log(level LogLevel, mcode MCode, message string, fields map[string]any) {
	if level < rcvr.level {
		return
	}
	entry := LogEntry{
		Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
		Level:     level.String(),
		Code:      mcode.Code,
		Component: rcvr.config.Component,
		Service:   rcvr.config.Service,
		Message:   optionalMessage(mcode, message),
		Fields:    cloneFields(fields),
	}
	if rcvr.config.EnableCaller || level == LevelDebug {
		entry.setCaller(3)
	}
	entry.extractError()
	rcvr.write(entry)
}

func cloneFields(fields map[string]any) map[string]any {
	if len(fields) == 0 {
		return nil
	}
	cloned := make(map[string]any, len(fields))
	for key, value := range fields {
		cloned[key] = value
	}
	return cloned
}

func (rcvr *LogEntry) setCaller(skip int) {
	pc, file, line, ok := runtime.Caller(skip)
	if !ok {
		return
	}
	rcvr.File = file
	rcvr.Line = line
	if fn := runtime.FuncForPC(pc); fn != nil {
		rcvr.Function = fn.Name()
	}
}

func (rcvr *LogEntry) extractError() {
	if rcvr.Fields == nil {
		return
	}
	switch value := rcvr.Fields["error"].(type) {
	case nil:
		return
	case error:
		rcvr.Error = value.Error()
	case string:
		rcvr.Error = value
	default:
		rcvr.Error = fmt.Sprint(value)
	}
	delete(rcvr.Fields, "error")
	if len(rcvr.Fields) == 0 {
		rcvr.Fields = nil
	}
}

func (rcvr *logger) write(entry LogEntry) {
	if rcvr.config.Structured {
		data, err := json.Marshal(entry)
		if err == nil {
			fmt.Fprintln(rcvr.output, string(data))
			return
		}
	}
	fmt.Fprintf(rcvr.output, "[%s] [%s] [%s] %s", entry.Timestamp, entry.Level, MCode{Code: entry.Code}.PaddedCode(), entry.Message)
	if len(entry.Fields) > 0 && entry.Level == LevelDebug.String() {
		if data, err := json.Marshal(entry.Fields); err == nil {
			fmt.Fprintf(rcvr.output, " %s", string(data))
		}
	}
	if entry.Error != "" {
		fmt.Fprintf(rcvr.output, " error=%q", entry.Error)
	}
	fmt.Fprintln(rcvr.output)
}

File: external/share/logger.ts
import { MCode } from './mcode';

export interface LoggerConfig {
  component?: string;
  service?: string;
  level?: string;
  structured?: boolean;
  enableCaller?: boolean;
  output?: string;
}

export interface Logger {
  DEBUG(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void;
  INFO(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void;
  WARN(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void;
  ERROR(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void;
  FATAL(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void;
}

export enum LogLevel {
  Debug = 0,
  Info = 1,
  Warn = 2,
  Error = 3,
  Fatal = 4,
}

export interface LogEntry {
  timestamp: string;
  level: string;
  code: string;
  component?: string;
  service?: string;
  message: string;
  fields?: Record<string, unknown>;
  file?: string;
  function?: string;
  line?: number;
  error?: string;
}

type LogWriter = (line: string) => void;

class SharedLogger implements Logger {
  private readonly level: LogLevel;
  private readonly output: LogWriter;

  constructor(private readonly config: LoggerConfig, output?: LogWriter) {
    this.level = parseLogLevel(config.level);
    this.output = output ?? openLogOutput(config.output);
  }

  DEBUG(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.Debug, mcode, optionalMessage, fields);
  }

  INFO(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.Info, mcode, optionalMessage, fields);
  }

  WARN(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.Warn, mcode, optionalMessage, fields);
  }

  ERROR(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.Error, mcode, optionalMessage, fields);
  }

  FATAL(mcode: MCode, optionalMessage: string, fields?: Record<string, unknown>): void {
    this.log(LogLevel.Fatal, mcode, optionalMessage, fields);
    if (typeof process !== 'undefined') process.exit(1);
    throw new Error(optionalMessageFor(mcode, optionalMessage));
  }

  private log(level: LogLevel, mcode: MCode, message: string, fields?: Record<string, unknown>): void {
    if (level < this.level) return;
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level: logLevelString(level),
      code: mcode.code,
      message: optionalMessageFor(mcode, message),
    };
    if (this.config.component) entry.component = this.config.component;
    if (this.config.service) entry.service = this.config.service;
    if (this.config.enableCaller || level === LogLevel.Debug) setCaller(entry);
    const clonedFields = cloneFields(fields);
    if (clonedFields) entry.fields = clonedFields;
    extractError(entry);
    this.write(entry);
  }

  private write(entry: LogEntry): void {
    if (this.config.structured) {
      this.output(JSON.stringify(entry));
      return;
    }
    let line = `[${entry.timestamp}] [${entry.level}] [${new MCode(entry.code, '').PaddedCode()}] ${entry.message}`;
    if (entry.fields && entry.level === logLevelString(LogLevel.Debug)) line += ` ${JSON.stringify(entry.fields)}`;
    if (entry.error) line += ` error=${JSON.stringify(entry.error)}`;
    this.output(line);
  }
}

export function NewLogger(config: LoggerConfig): Logger {
  return newLogger(config);
}

export function NewEnvLogger(component: string, service: string): Logger {
  const env = processEnv();
  const config: LoggerConfig = {
    component,
    service,
    structured: truthyEnv(env.XALIGO_LOG_STRUCTURED),
    enableCaller: truthyEnv(env.XALIGO_LOG_CALLER),
  };
  if (env.XALIGO_LOG_LEVEL !== undefined) config.level = env.XALIGO_LOG_LEVEL;
  if (env.XALIGO_LOG_OUTPUT !== undefined) config.output = env.XALIGO_LOG_OUTPUT;
  return NewLogger(config);
}

export function newLogger(config: LoggerConfig, output?: LogWriter): Logger {
  return new SharedLogger(config, output);
}

function processEnv(): Record<string, string | undefined> {
  if (typeof process === 'undefined') return {};
  return process.env;
}

function truthyEnv(value: string | undefined): boolean {
  switch ((value ?? '').trim().toLowerCase()) {
    case '1':
    case 'true':
    case 'yes':
    case 'y':
    case 'on':
      return true;
    default:
      return false;
  }
}

function parseLogLevel(level: string | undefined): LogLevel {
  switch ((level ?? '').trim().toUpperCase()) {
    case 'DEBUG':
      return LogLevel.Debug;
    case 'WARN':
    case 'WARNING':
      return LogLevel.Warn;
    case 'ERROR':
      return LogLevel.Error;
    case 'FATAL':
      return LogLevel.Fatal;
    default:
      return LogLevel.Info;
  }
}

function logLevelString(level: LogLevel): string {
  switch (level) {
    case LogLevel.Debug:
      return 'DEBUG';
    case LogLevel.Info:
      return 'INFO';
    case LogLevel.Warn:
      return 'WARN';
    case LogLevel.Error:
      return 'ERROR';
    case LogLevel.Fatal:
      return 'FATAL';
    default:
      return 'UNKNOWN';
  }
}

function openLogOutput(output: string | undefined): LogWriter {
  const target = (output ?? '').trim();
  switch (target) {
    case 'stdout':
      return (line: string) => console.log(line);
    case '':
    case 'stderr':
      return (line: string) => console.error(line);
    default:
      return (line: string) => appendLogFile(target, line);
  }
}

function appendLogFile(path: string, line: string): void {
  if (typeof require !== 'undefined') {
    const fs = require('node:fs') as { appendFileSync(path: string, data: string): void };
    fs.appendFileSync(path, `${line}\n`);
    return;
  }
  console.error(`logger output fallback: ${path}: file output is unavailable`);
  console.log(line);
}

function optionalMessageFor(mcode: MCode, optional: string): string {
  if (optional === '') return mcode.message;
  if (mcode.message === '') return optional;
  return `${mcode.message}: ${optional}`;
}

function cloneFields(fields: Record<string, unknown> | undefined): Record<string, unknown> | undefined {
  if (!fields || Object.keys(fields).length === 0) return undefined;
  return { ...fields };
}

function extractError(entry: LogEntry): void {
  if (!entry.fields || !Object.prototype.hasOwnProperty.call(entry.fields, 'error')) return;
  const value = entry.fields.error;
  if (value instanceof Error) entry.error = value.message;
  else if (typeof value === 'string') entry.error = value;
  else if (value !== undefined && value !== null) entry.error = String(value);
  delete entry.fields.error;
  if (Object.keys(entry.fields).length === 0) delete entry.fields;
}

function setCaller(entry: LogEntry): void {
  const stack = new Error().stack?.split('\n').map((line) => line.trim()) ?? [];
  const caller = stack.find((line) => !line.includes('SharedLogger.') && !line.includes('setCaller'));
  if (!caller) return;
  entry.function = caller;
}